Linear constraints proposal (#621)
ghc-proposals/ghc-proposals: proposals/0621-linear-constraints.rst

When ``-XLinearTypes`` is enabled, the following new syntax is
enabled:

::

   ctype ::= context '%' 1 '=>' ctype | context '=>' ctype | type | ...

Note that unlike multiplicities for function type arrows, linear
constraint arrows may only be instantiated with a ``1`` (linear)
multiplicity. Unlike linear function types, this proposal does not
introduce multiplicity polymorphism in constraint arrows, so there is
never a need for the multiplicity to be anything other than a ``1`` -
an ``Many`` multiplicity is already represented by omitting the
multiplicity entirely (i.e. using a standard constraint). Standard
non-linear constraints can then still be used in combination with
linear ones, with the order of linear and unrestricted constraints not
mattering.

Semantic/Desugaring
^^^^^^^^^^^^^^^^^^^

As typical, we define the semantics of linear constraints via
desugaring. The linear fat arrow ``%1 =>`` desugars to a linear arrow
``%1 ->``. Namely, writing ``⦇·⦈`` for the desugaring function:

- ``⦇C %1 => A⦈ = C %1 -> ⦇A⦈``
- ``⦇e :: C %1 => A⦈ = \(%1 $d :: C) -> ⦇e :: A⦈``

This desugaring means that changes to GHC Core itself are not

> required: we only need the material introduced by linear types and
> already implemented.

Reviewer: ```suggestion
required: we only need the material introduced by linear types which has
already been implemented in GHC 9.0.
```

one method, the one method is linear in the class constraint

::

  -- Given
  class Bar a where
    h :: H a

  -- We have
  h :: Bar a %1 => H a

Typing inference
^^^^^^^^^^^^^^^^

In a way, there's no need to worry about type inference: if a function
``C %1 -> T``, with well-placed dictionaries, would be rejected, then
``C %1 => T`` will be rejected as well. So understanding linear types
is sufficient for the most part.

But when there is an accepted assignment of type ``C %1 -> T``, it
doesn't follow that the function of type ``C %1 => T`` will be
accepted. Because GHC's typechecker doesn't make guesses.

The one new rule introduced by this proposal is that when I want a
linear constraint ``C`` and I've been given both a linear an an
unrestricted ``C``, then this is considered ambiguous and raises a
type error. See Section 6.3 of the paper_ for more details.

To see why, consider this example

::

  class C
  giveC :: (C => Int) -> Int
  useC :: C %1 => Int

  bad :: C %1 => (Int, Int)
  bad = (giveC useC, useC)

  bad' :: C %1 => (Int, Int)
  bad' = (giveC useC, 0)

In ``bad``, if the leftmost ``useC`` uses the linear ``C`` from the

> function signaturev, then ``bad`` would be rejected, it must used the

Reviewer: ```suggestion
function signature, then ``bad`` would be rejected, it must used the
```

Given equality constraints are used for unification *only
if they are unrestricted* (correspondingly, equality constraints
generated by the type inference algorithm are unrestricted, as they
have always been).

The reason for this is that there is no clear semantics to make use of
a linear equality constraint ``a ~ b`` as part of the unification
algorithm (it's not that reasoning about linear equality is
meaningless, but a unification or congruence conversion for linear
equality isn't obvious to come up with, if someone has we're not
aware). Nor does it feel like a true limitation as there is no example
where a linear equality would be useful. It's really not worth the
bother of trying to find a solution.

Dupable classes
^^^^^^^^^^^^^^^

A new module ``GHC.Constraint.Linear`` is introduced (inlined
alternative: bikeshed names, including the module name).

This module exposes the following:

::

  data LinearityToken :: ZeroBitType

  consumeLinearityToken :: LinearityToken %1 -> (# #)
  dup2LinearityToken :: LinearityToken %1 -> (# LinearityToken, LinearityToken #)

  data DupableClassModifier = Dupable

Class declaration can be annotated with the ``%Dupable``

::

>   class %Dupable <ctx> => <head> where

Reviewer: We might be at odds here with the modifiers GHC proposal

Author: Did I place the modifier in the wrong place? I need to double-check.

    <methods>

> Classes annotated with ``%Dupable`` must:
> - Have a single method
> - The method must be of type ``LinearityToken``

Reviewer: ```suggestion
Classes annotated with ``%Dupable`` must:

- Have a single method
- The method must be of type ``LinearityToken``
```

::

  read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
  read2AndDiscard arr =
       let !(Box x)  = read arr 0
           !(Box y)  = read arr 1
           !()       = free arr
       (x, y)

  linearly $
    let
      !(NewMArray arr1) = new
      !(NewMArray arr2) = new
    in
    … -- modify the array as suited
    Ur $ sum arr1 + sum arr2

There is still a little bit of noise there, what with the defining of
GADTs (while ``/\`` can be defined once and for all, types like
``NewMArray`` must be defined for most every type because we lack
type-level lambdas), and the constructors in the let-bindings.

With existential types, this would look something like this (the
existential types proposal defines, not coincidentally, a ``/\`` with
the same role as that above):

::

  read  :: Read n %1 => MArray a n -> Int -> Read n /\ a
  free :: (Read n, Write n) %1 => MArray a n -> ()

  new :: Linearly %1 => Int -> exists n. (Read n, Write n) /\ MArray a n

  read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
  read2AndDiscard arr =
       let !x  = read arr 0
           !y  = read arr 1
           !() = free arr

>        (x, y)

Reviewer 2: I think we need this here?
```suggestion
       (Ur x, Ur y)
```

Author: Actually, I meant for `read` to return an `Ur a`, I'll fix soon.

never a need for the multiplicity to be anything other than a ``1`` -
an ``Many`` multiplicity is already represented by omitting the
multiplicity entirely (i.e. using a standard constraint). Standard
non-linear constraints can then still be used in combination with
linear ones, with the order of linear and unrestricted constraints not
mattering.

Semantic/Desugaring
^^^^^^^^^^^^^^^^^^^

As typical, we define the semantics of linear constraints via
desugaring. The linear fat arrow ``%1 =>`` desugars to a linear arrow
``%1 ->``. Namely, writing ``⦇·⦈`` for the desugaring function:

- ``⦇C %1 => A⦈ = C %1 -> ⦇A⦈``
- ``⦇e :: C %1 => A⦈ = \(%1 $d :: C) -> ⦇e :: A⦈``

This desugaring means that changes to GHC Core itself are not
required: we only need the material introduced by linear types and
already implemented.

Using type classes linearly
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Type class methods require an unrestricted class constraint:

::

  -- Given
  class Foo a where
    f :: F a
    g :: G a

  -- We have
  f :: Foo a => F a
  g :: Foo a => G a

This is unchanged. But we add one exception: in type classes with exactly

> one method, the one method is linear in the class constraint

Reviewer 3: I don't really understand the motivation for this. I get a sense it is justified by the need to use the `Linearly` in a linear way but it seems rather vague to me. Perhaps it could be spelled out more fully.

Author: I suppose this needs an explanation somewhere. It's not particularly related to the `Linearly` type class. Rather, it's to address that we need at least one function of type `C %1 => T`, otherwise we couldn't in any circumstance have a term of type `C %1 => U` (they would fail either because the `C` is used in a function of type `C =>`, or not used at all, neither of which is permitted).

    <methods>

Classes annotated with ``%Dupable`` must:

- Have a single method
- The method must be of type ``LinearityToken``

Such a dupable type class can be used multiple times even if they are
linear. *E.g.*::

  class Foo where
    foo' :: LinearityToken

  foo :: Foo %1 => Int -> Int
  foo = case foo' of { (# #) -> \x -> x }

  dupes :: Foo %1 => (Int -> Int, Int -> Int)
  dupes = (foo, foo)

But, crucially, not passed to an unrestricted function::

  rejected :: Foo %1 => Ur (Int -> Int)
  rejected = Ur foo

There are a lot of alternatives for the design of this feature, so see
the *Alternatives* section for more thoughts.

Examples
--------

Threading
^^^^^^^^^

We refer back now to the first example from the motivation section,
which showed how writing a function which reads the first two elements
of an array became a tedious exercise of threading our linear resource
through the function. Using linear constraints, however, such a
function can be written as:

::

   read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
   read2AndDiscard arr =
        let !(Box x)  = read arr 0
            !(Box y)  = read arr 1
            !()       = free arr

>         (x, y)

Reviewer 4: ```suggestion
        in (x, y)
```

management for linear logic proof search`_, where such higher-order
givens are handled. The extension is unproblematic, it would
presumably be more effort to prevent it than to support it.

Therefore, when ``-XLinearTypes`` and ``-XLinearConstraints`` are both
on, contexts can contain implications of the form ``C %1 => D``.

Existential types proposal
^^^^^^^^^^^^^^^^^^^^^^^^^^

The `Existential types proposal`_, if they ever materialise (🤞), will
make this proposal even better. Using linear constraints in APIs most
often require returning constraints too. For that we've been using
GADTs. This was our simple mutable array API:

::

  read  :: Read n %1 => MArray a n -> Int -> Read n /\ a

  free :: (Read n, Write n) %1 => MArray a n -> ()

  data c /\ a where
    Box :: c %1 => a -> c /\ a

  new :: Linearly %1 => Int -> NewMArray a

  data NewMArray a where
    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a n

With this API we can write functions such as

::

  read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
  read2AndDiscard arr =
       let !(Box x)  = read arr 0
           !(Box y)  = read arr 1
           !()       = free arr

>        (x, y)

Reviewer 3: I don't understand if this is part of a proposed change to strict bindings. This wouldn't work in current GHC, would it? For example, in the below only `showItCase`  works, `showItLet` doesn't.

```haskell
data WithShow a where
  MkWithShow :: Show a => a -> WithShow a

showItCase :: WithShow a -> String
showItCase x = case x of MkWithShow a -> show a

showItLet :: WithShow a -> String
showItLet x = let !(MkWithShow a) = x in show a
```

Author: No, I just didn't realise it. I guess we'd done all of our examples with `case`.

I understand why it's so: these things desugar to a form that isn't conducive to manipulating existential types.

In fact, in the [linear let MR](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10952), I'm changing the desugaring of this very case (strict, non-recursive, monomorphic pattern bind) to be a regular `case`. We could change the type checker to understand that such strict, non-recursive, monomorphic pattern binds can bind type variables and constraint evidence, but maybe it's a little fragile?

the first two elements, and then de-allocates it before returning the
two elements. The fact that the array is a linear resource provides
some nice guarantees which ensure memory safety. For example, there is
no way we can go on to erroneously read from the array after
de-allocating it. Instead, when ``arr0`` is used as an argument to
``read``, ``read`` consumes it and returns the array under a new name
``arr1``. This process then repeats before finally ``free``
de-allocates the array.

While this code ensures that we use the ``MArray`` in a memory-safe
way, this re-naming process introduces boilerplate and becomes
cumbersome to both write and read. Compare to the doing the same with
the ``ST`` monad instead:

::

   read2AndDiscardST :: MArray s a -> ST s (a, a)
   read2andDiscardST arr = do
     x <- read arr 0
     y <- read arr 1
     return (x, y)

There is much less noise. To be honest, this is all a little silly:
the name management is completely mechanical; it should be handled
automatically by the compiler. This is what this proposal lets us do:
get pretty close to the latter style, without having to introduce a

> monad.

Author: I'm taking this kind of for granted, because linear constraints or not, the whole point of this array example is to avoid monads. The simple fact that they are considered appealing by anyone (in fact it's the first example that people think of when they think of linear types, and gets them excited (though it's always been my position that it's not the most compelling example)) is evidence that mutation without monad is attractive.

That being said, the motivation section does link to [this blog post of mine](https://www.tweag.io/blog/2023-01-26-linear-constraints-freeze/) that gives you an example (recursive freezing) that seems absolute madness without linear constraints (even in the ST monad!).

> monad.

Reviewer 5: One more cool thing that is possible with linear constraints and isn't possible with vanilla IO/ST monad is to control that we don't use mutable buffer after deallocation/handle after closing or something similar.

A couple of months ago, I caught a bug with vanilla file handle API when I attempt to write into handle opened into ReadWrite mode after `hGetContents` function and got exception about closed handle. 

```haskell
withFile "file-name" ReadWriteMode \h -> do
  content <- Bytestring.Strict.hGetContents h
  hPutStr h "write into handle"
  pure content
```
It is turned out that `hGetContents` closes handle at the end.

With linearity-based API `hGetContents` just would not return handle Read/Write constraints, so bad code just would not compile.

continuation as an argument

::

   new :: Int -> (MArray %1 -> Ur a) %1 -> Ur a

This forces the array to be single-threaded (thanks to the ``Ur a``
return type, the ``MArray`` cannot escape the continuation's scope),
which we can use to guarantee uniqueness.

This is a little clumsy to program with. But more importantly, these
continuations aren't very composable as argued in `this blog post
<blog_scopes_>`_. See also the long discussion at
`tweag/linear-base#130
<https://github.com/tweag/linear-base/issues/130>`_. This proposal
will let us define ``new`` in direct style.

Proposed Change Specification
-----------------------------

Main Syntax Changes
^^^^^^^^^^^^^^^^^^^

Currently, type class constraints in GHC do not support multiplicty
annotations.  GHC currently defines the syntax for type signatures as:

::

   ctype   ::= context '=>' ctype | type | ...

Essentially, type signatures can consist of (among other things which
we ignore here) zero or more qualified type arrows ``=>`` followed by
a type. Here ``context`` is a list of class constraints.

When ``-XLinearTypes`` is enabled, the following new syntax is
enabled:

::

>    ctype ::= context '%' 1 '=>' ctype | context '=>' ctype | type | ...

Reviewer 6: But what if I want to have one linear and one ordinary constraint with multiplicity "`Many`"? What if I write:

```haskell
foo :: (C, Integral a) %1 => a %1 -> a %1 -> Integer`
foo a1 a2 = toInteger a1 + toInteger a2 + toInteger useC
```
I've used each `a` argument once, `C` once, but `Integral a` twice. But I think this would be forbidden by your proposal, and the only way to support it is to change the context to `(C, Integral a, Integral a) %1`. You say later that a combination is possible, but I find it hard to understand how. Maybe `C %1 => Integral a => a %1 -> a %1 -> Integer` is allowed by your syntax? That looks strange to me.

It is strange, in my opinion, that the `%1` is attached to the whole constraint context and not just a single type class. E.g. I'd expect `(C %1, C %1) => Int` instead of `(C, C) %1 => Int`. If this was the proposed syntax, then I could also combine linear and normal constraint with `(C %1, OtherConstraint) => ...`, and use `OtherConstraint` as often as I want.

Maybe I'm also misunderstanding something here, but that might then be a sign that this case could be clearer ;)

Author: Under this proposal, you'd write

```haskell
foo :: C %1 => Integral a => a %1 -> a %1 -> Integer
```

There is no theoretical obstacle here. I proposed the most minimalistic syntax change, largely because I'm scared of GHC's parser. There's also no support for unnamed tuples with fields of varying multiplicity in Core, and I believe (though I haven't checked) that these tuples of constraints are desugared as tuples. Anything that doesn't change Core makes my life much easier.

See [_Unrestricted modifier for constraints_](https://github.com/tweag/ghc-proposals/blob/linear-constraints/proposals/0621-linear-constraints.rst#74unrestricted-modifier-for-constraints) in the _Alternatives_ section.

required: we only need the material introduced by linear types, which
is part of GHC since GHC 9.0.

Using type classes linearly
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Type class methods require an unrestricted class constraint:

::

  -- Given
  class Foo a where
    f :: F a
    g :: G a

  -- We have
  f :: Foo a => F a
  g :: Foo a => G a

This is unchanged. But we add one exception: in type classes with exactly
one method, the one method is linear in the class constraint
(otherwise there would never be inhabitants in the type ``C %1 => T``)

::

  -- Given
  class Bar a where
    h :: H a

  -- We have
  h :: Bar a %1 => H a

Typing inference
^^^^^^^^^^^^^^^^

In a way, there's no need to worry about type inference: if a function
``C %1 -> T``, with well-placed dictionaries, would be rejected, then
``C %1 => T`` will be rejected as well. So understanding linear types
is sufficient for the most part.

But when there is an accepted assignment of type ``C %1 -> T``, it
doesn't follow that the function of type ``C %1 => T`` will be
accepted. Because GHC's typechecker doesn't make guesses.

The one new rule introduced by this proposal is that when I want a

> linear constraint ``C`` and I've been given both a linear an an

Reviewer 6: ```suggestion
linear constraint ``C`` and I've been given both a linear and an
```

   %Dupable class <ctx> => <head> where
    <methods>

Classes annotated with ``%Dupable`` must:

- Have a single method
- The method must be of type ``LinearityToken``

Such a dupable type class can be used multiple times even if they are
linear. *E.g.*::

  class Foo where
    foo' :: LinearityToken

  foo :: Foo %1 => Int -> Int
  foo = case foo' of { (# #) -> \x -> x }

  dupes :: Foo %1 => (Int -> Int, Int -> Int)
  dupes = (foo, foo)

But, crucially, not passed to an unrestricted function::

  rejected :: Foo %1 => Ur (Int -> Int)
  rejected = Ur foo

There are a lot of alternatives for the design of this feature, so see
the *Alternatives* section for more thoughts.

Examples
--------

Threading
^^^^^^^^^

We refer back now to the first example from the motivation section,
which showed how writing a function which reads the first two elements
of an array became a tedious exercise of threading our linear resource
through the function. Using linear constraints, however, such a
function can be written as:

::

   read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
   read2AndDiscard arr =
        let !(Box x)  = read arr 0
            !(Box y)  = read arr 1

>             !()       = free arr

Reviewer 3: Yes, I wonder if it's OK to use `do` notation instead, or whether that defeats the point of the proposal.

>             !()       = free arr

Reviewer 5: It is possible to express so-called sequenced "let no rec" using qualified do notation with preserve of linearity: https://play.haskell.org/saved/5ahwqWY2

However, I don't think that this is a good idea to introduce do-notation based DSL in the proposal and then write all examples in terms of this DSL.

>             !()       = free arr

Author: @JakobBruenker 

> Here, I assume that order is determined by the order of bindings within the `let` block? Which makes sense, but is a bit surprising, since AFAIK until now, order of bindings within a `let` block has not been significant.

You can observe the order of strict bindings today

```haskell
let {!x = error "1"; !y = error "2"} in 0
```

GHC is rather careful around strict bindings. Though I can't think of a way to make it matter (that is, returning a different exception is considered to be the same semantics in Haskell; and I don't know whether we can concoct an example where the order of evaluation of let arguments will say, turn an exception into a non-exception).

The way the type-checker is specified in the paper also makes it so that there's a single solution (in Core, the ordering is simply the threading of a state token, so only the type-checker matters).

So you take away what you want from this: it's a possibly surprising twist on something that already exists.

>             !()       = free arr

Reviewer 2: Interesting! Seems like it would be useful to spell out explicitly in the proposal that this is how the order is determined.

>             !()       = free arr

Reviewer 2: (I think github will present this without context, so to be clear, this is a response in [an existing discussion thread](https://github.com/ghc-proposals/ghc-proposals/pull/621/files#r1412909224))

The example you provide here
```haskell
let {!x = error "1"; !y = error "2"} in 0
```
produces
```
*** Exception: 2
```
Doesn't that mean strict `let` bindings have the opposite evaluation order from what we would want?

(NB, the syntax I suggest in a different comment has the order one would expect:)

```haskell
ghci> if | !x <- error "1", !y <- error "2" -> ()
*** Exception: 1
```

   consuming n = useC + n

The constraint ``C`` is consumed exactly once (namely by the call to
``useC``).

On the other hand, the following program is rejected:

::

   neglecting :: C %1 => Int -> Int
   neglecting n = n + n

as ``C`` is never consumed.

Furthermore, as with linear arguments, we must *guarantee* that ``C``
will be consumed, regardless of the conditional branching. Thus, the
following would also be rejected:

::

   dithering :: C %1 => Bool -> Int
   dithering x = if x then useC else 10
as ``useC`` is only consumed when ``x == True``.

These examples show programs rejected by failing to guarantee that
``C`` will be used. However, as we are in a linear context, we must
also ensure that it isn't *overused*:

::

   overusing :: C %1 => (Int, Int)
   overusing = (useC, useC)

Since ``overusing`` consumes a linear ``C`` constraint twice, it is
rejected for violating linearity. We may amend the type scheme of
``overusing`` in the following way, however:

::

   stammering :: (C, C) %1 => (Int, Int)
   stammering = (useC, useC)

By providing an additional linear constraint, the resource usage
guarantees are satisfied and the program type checks.

> The theoretical basisof this proposal, as well as further examples,

Reviewer 7: ```suggestion
The theoretical basis of this proposal, as well as further examples,
```

to identify the array. Accordingly, our type constructor for
``MArray`` is parameterised by ``n``.

The type signatures for  ``read``, ``free``, and ``Box`` are:

::

   read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a

   free :: (Read n, Write n) %1 => MArray a n -> ()

   data c /\ a where
     Box :: c %1 => a -> c /\ a

i.e. ``read`` is a function which consumes a linear ``Read n``
constraint, allowing us to read from the specified array index. It also
returns a new ``Read n`` constraint, allowing us to subsequently read
from the array again. Likewise, ``free`` consumes both a ``Read n``
and a ``Write n`` constraint and introduces none, ensuring that we
cannot read or write after freeing.

Thus we eliminate the need to manually thread the ownership of the
array through the function, whilst maintaining the guarantees of
unique ownership via the linear constraints.

For a more in-depth example along these lines, refer to section 4 of
the paper_.

The Linearly constraint
^^^^^^^^^^^^^^^^^^^^^^^

We can create a class, the paper_ calls it ``Linearly`` with the
following API:

::

  %Dupable class Linearly

>   linearly :: (Linearly => Ur a) %1 -> Ur a

Reviewer 8: Just a minor nitpick, but shouldn't this be:

```hs
 linearly :: (Linearly %1=> Ur a) %1 -> Ur a
```

Quantified Constraints
^^^^^^^^^^^^^^^^^^^^^^

Our syntax extension naturally extends constraints in types' context
to support linear implications ``C %1 => D`` when
``-XQuantifiedConstraint`` is on. While this is not described in the
paper_, our solving algorithm is based on `Efficient resource
management for linear logic proof search`_, where such higher-order
givens are handled. The extension is unproblematic, it would
presumably be more effort to prevent it than to support it.

Therefore, when ``-XLinearTypes`` and ``-XLinearConstraints`` are both
on, contexts can contain implications of the form ``C %1 => D``.

Existential types proposal
^^^^^^^^^^^^^^^^^^^^^^^^^^

The `Existential types proposal`_, if they ever materialise (🤞), will
make this proposal even better. Using linear constraints in APIs most
often require returning constraints too. For that we've been using
GADTs. This was our simple mutable array API:

::

  read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a

  free :: (Read n, Write n) %1 => MArray a n -> ()

  data c /\ a where
    Box :: c %1 => a -> c /\ a

  new :: Linearly %1 => Int -> NewMArray a

  data NewMArray a where

>     NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a n

Reviewer 5: ```suggestion
    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a
```

``Write n`` linear constraints. Here, ``n`` is a type-level tag used
to identify the array. Accordingly, our type constructor for
``MArray`` is parameterised by ``n``.

The type signatures for  ``read``, ``free``, and ``Box`` are:

::

   read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a

   free :: (Read n, Write n) %1 => MArray a n -> ()

   data c /\ a where
     Box :: c %1 => a -> c /\ a

i.e. ``read`` is a function which consumes a linear ``Read n``
constraint, allowing us to read from the specified array index. It also
returns a new ``Read n`` constraint, allowing us to subsequently read
from the array again. Likewise, ``free`` consumes both a ``Read n``
and a ``Write n`` constraint and introduces none, ensuring that we
cannot read or write after freeing.

Thus we eliminate the need to manually thread the ownership of the
array through the function, whilst maintaining the guarantees of
unique ownership via the linear constraints.

For a more in-depth example along these lines, refer to section 4 of
the paper_.

The Linearly constraint
^^^^^^^^^^^^^^^^^^^^^^^

We can create a class, the paper_ calls it ``Linearly`` with the
following API:

::

>   %Dupable class Linearly

Reviewer 2: If, as mentioned below, all `%Dupable` classes are isomorphic, wouldn't it make sense to also include that class in `GHC.Constraint.Linear`?

  linearly :: (Linearly %1 => Ur a) %1 -> Ur a
  newLinearlyDict :: Linear.IO (Dict Linearly)

  data Dict c where
    Dict :: c %1 => Dict c

This ensures that it is not possible to ever build an unrestricted
evidence for ``Linearly``.

Equipped with this we can extend the API of the example above with
a way to create arrays::

  new :: Linearly %1 => Int -> NewMArray a

  data NewMArray a where
    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a n

Because there is no unrestricted evidence of ``Linearly``, the
linearity of the ``Linearly`` constraint will contaminate the returned
``NewMArray a`` value, ensuring in turn that the returned ``Read n``,
and ``Write n`` constraints *must* be used linearly, as required.

The difference with having ``new`` itself use a continuation is that
we can now have several calls to ``new`` in the same scope. Which
prevents the problems described in the *Motivation* section.

::

  linearly $
    let
      !(NewMArray arr1) = new
      !(NewMArray arr2) = new
    in
    … -- modify the array as suited

>     Ur $ sum arr1 + sum arr2

Reviewer 2: It would be good to explain somewhere (perhaps in the context of this example) why `Linearly` needs to be `Dupable`, as the paper does

make this proposal even better. Using linear constraints in APIs most
often require returning constraints too. For that we've been using
GADTs. This was our simple mutable array API:

::

  read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a

  free :: (Read n, Write n) %1 => MArray a n -> ()

  data c /\ a where
    Box :: c %1 => a -> c /\ a

  new :: Linearly %1 => Int -> NewMArray a

  data NewMArray a where
    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a n

With this API we can write functions such as

::

  read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
  read2AndDiscard arr =
       let !(Box x)  = read arr 0
           !(Box y)  = read arr 1
           !()       = free arr
       in (x, y)

  linearly $
    let
      !(NewMArray arr1) = new
      !(NewMArray arr2) = new
    in
    … -- modify the array as suited
    Ur $ sum arr1 + sum arr2

There is still a little bit of noise there, what with the defining of
GADTs (while ``/\`` can be defined once and for all, types like
``NewMArray`` must be defined for most every type because we lack
type-level lambdas), and the constructors in the let-bindings. Plus

> this is cheating a tad: let bindings, even strict lets, don't

Reviewer 5: As linear constraints have the same rules as linear functions, it is possible to try out such ideas using exising libraries and manual token passing.

WRT your particular example, current GHC always consider pattern guards non-linear consumption (https://play.haskell.org/saved/wvGeLDMJ)

Even with linear let it is not possible to consume linear value. This example works:
```haskell
read2AndDiscard :: Array a %1 -> Ur (a, a)
read2AndDiscard arr
  = let
  !(Ur x, arr') = read arr 0
  !(Ur y, arr'') = read arr' 
  in case consume arr'' of
    () -> Ur (x, y)
```
And this doesn't:
```haskell
read2AndDiscard :: Array a %1 -> Ur (a, a)
read2AndDiscard arr
  = let
  !(Ur x, arr') = read arr 0
  !(Ur y, arr'') = read arr' 
  !() = consume arr''
  in Ur (x, y)
```

Explicit bangs and consuming array in case do not work with pattern guards either.

> this is cheating a tad: let bindings, even strict lets, don't

Author: The second example is fixed in master. But since it was discovered after the 9.12 fork and is not a soundness bug, it'll only be possible to write it in 9.14.

proposal without any dupable type classes. If the design of this
feature proves too controversial, it's definitely an option to simply
excise the dupable class feature from the proposal.

Kind of LinearityToken
~~~~~~~~~~~~~~~~~~~~~~

The proposal deliberately specifies that ``LinearityToken`` be 0
width, and the ``dup2`` and ``consume`` functions to correspondingly
return unboxed tuples. This is meant to emphasise that this is all for
low level manipulation and making sure that there is no cost in
storing ``LinearityToken``. These are all meant to be used while
defining a dupable type class and its API, but it's not intended for
them to be apparent in said API.

An somewhat middle-ground option is to expose

::

  data LinearityToken# :: ZeroBitType

  data LinearityToken = MkLinearityToken LinearlyToken#

  -- Both functions below can be defined as easily inlineable thin
  -- wrapper so that in most cases no allocation is needed
  consumeLinearityToken :: LinearityToken -> ()
  dup2LinearityToken :: LinearityToken -> (LinearityToken, LinearityToken)

Just linearly
~~~~~~~~~~~~~

The most useful dupable type class in the proximate future (in fact

> the only known example yet), is ``Linearly``. So another option, to

Reviewer 2: To be clear, the fact that all dupable type classes are isomorphic means it's the only possible example, doesn't it?

avoid introducing any ad hoc syntax is to simply expose the (abstract)
``Linearly`` constraint from the ``GHC.Constraint.Linear`` module *and
nothing else*. So that ``GHC.Constraint.Linear`` would be

::

  module GHC.Constraint.Linear where

  -- Magically dupable
  class Linearly

  linearly'sToken :: Linearly %1 => LinearityToken

  data LinearityToken :: ZeroBitType

  consumeLinearityToken :: LinearityToken -> (# #)
  dup2LinearityToken :: LinearityToken -> (# LinearityToken, LinearityToken #)

This is quite economical from a language extension perspective, but
the authors of this proposal are somewhat worried of the difficulties

> of wiring in a type class.

Reviewer 2: To me, this alternative seems like the best design, especially given its forward-compability. However, given the potential difficulties in wiring in a type-class, I'm willing to go along with the current design, too.

  bad :: C %1 => (Int, Int)
  bad = (giveC useC, useC)

  bad' :: C %1 => (Int, Int)
  bad' = (giveC useC, 0)

In ``bad``, if the leftmost ``useC`` uses the linear ``C`` from the
function signature, then ``bad`` would be rejected, it must used the
unrestricted ``C`` from ``giveC``. But in ``bad'`` it must use the
linear ``C`` instead. So this would force the leftmost ``useC`` to
make a guess. Instead we reject both ``bad`` and ``bad'``.

Equality constraints
~~~~~~~~~~~~~~~~~~~~

Given equality constraints are used for rewriting *only
if they are unrestricted* (correspondingly, equality constraints
generated by the type inference algorithm are unrestricted, as they
have always been).

The reason for this is that there is no clear semantics to make use of
a linear equality constraint ``a ~ b`` as part of the unification
algorithm (it's not that reasoning about linear equality is
meaningless, but a unification or congruence conversion for linear
equality isn't obvious to come up with, if someone has, we're not
aware). Nor does it feel like a true limitation as there is no example
where a linear equality would be useful. It's really not worth the
bother of trying to find a solution.

> Dupable classes

Author: I can't prove that it isn't possible. But I don't know how to do it, and it's presumably quite challenging. Though I do agree that it would be nice.

The first observation is that the solver doesn't support axioms or hypotheses of the form `C1 => (C2, C3)`, the reason is presumably that this potentially is a big source of non-determinism. If I'm trying to solve `C2` and I use an hypothesis of this form I have, then, to prove `(C2 ⇒ C2)∨(C3 ⇒ C2)`. If `C3 ⇒ C2` is provable then we have two different proofs of `C2`. Which one do we use? GHC doesn't want to do this sort of choice.

And it gets worse with linear types, because what I wrote above is the rule for the additive product, but we want the multiplicative product for our use-case. So, instead, I have to retroactively use `C3` in the proof of `C2`. Blech…

This is why, in the paper, the dupable class have a special treatment, because it's a constrained usage of product which we can in fact implement. But this seems to require that the knowledge that a constraint is dupable or not is known a priori, so if we need the solver to give us this information, we need to run a separate instance of the solver (on each candidate constraint), it can be costly.

Yet another issue, is that this staging implies that the class is wired in, and you probably can't express this class without using `Dict`. This means wiring in a GADT, and using it during evidence construction. This is at least awkward, potentially very costly to implement (but I don't know, I haven't tried).

   ma >>= k = Linear.do { Box a <- ma; k a }

   (>>) :: m (c /\ ()) %1 -> (c %1 => m b) %1 -> m b
   mu >> mb = Linear.do { Box () <- mu; mb }

We'd have to change the APIs to always return a ``c /\ a`` even if
``c=()``. E.g.

::

   free :: (Read n, Write n) %1 => MArray a n -> () /\ ()

The code samples in this proposal use explicit ``Box``-s for clarity,
but it isn't meant to be prescriptive, we'll need more experience to
decide which style is best (it's worth pointing out that the
difference vanishes under the existential types proposal, see below).

The Linearly constraint
^^^^^^^^^^^^^^^^^^^^^^^

We can create a class, the paper_ calls it ``Linearly`` with the
following API:

::

  %Dupable class Linearly

  linearly :: (Linearly %1 => Ur a) %1 -> Ur a
  newLinearlyDict :: Linear.IO (Dict Linearly)

  data Dict c where
    Dict :: c %1 => Dict c

This ensures that it is not possible to ever build an unrestricted
evidence for ``Linearly``.

Equipped with this we can extend the API of the example above with
a way to create arrays::

  new :: Linearly %1 => Int -> NewMArray a

  data NewMArray a where
    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a

> 

Author: Presumably, I'd define a particular type, so something like

```haskell
class Read n where
  getReadPermission :: ReadPermission

data ReadPermission = ReadPermission
```

Maybe it's overkill, but I'd start there. Of course, the method musn't be exported.

You can then implement `read` as

```haskell
read :: Read n %1 => MArray n a -> Read n /\ a
read arr = case getReadPermission of
  ReadPermission -> unsafePerformSomethingLikeIO $
    r <- … -- actually read the array here
    <create a new instance of Read n by type-class reflection>
   Box r
```

It does get a little involved as an example, I don't know if I should elaborate in the proposal.

> 

Reviewer 9: I don't know what the commitee's opinion is but I'm happy to just know that
- Read is a linearitytoken# class
- Type-class reflection is used to create the instance

(And I can infer Write is similar.)
The latter should really be part of the proposal though, since you'll have to change withDict or add a new primop.

> 

Reviewer 5: > Read is a linearitytoken# class

But it's not

> 

Reviewer 9: You're right I read the wrong example, I should have said that it's a single method class

Because there is no unrestricted evidence of ``Linearly``, the
linearity of the ``Linearly`` constraint will contaminate the returned
``NewMArray a`` value, ensuring in turn that the returned ``Read n``,
and ``Write n`` constraints *must* be used linearly, as required.

The difference with having ``new`` itself use a continuation is that
we can now have several calls to ``new`` in the same scope. Which
prevents the problems described in the *Motivation* section.

::

  linearly $ DataFlow.do
    (NewMArray arr1) <- new
    (NewMArray arr2) <- new
    … -- modify the array as suited

>     Ur $ sum arr1 + sum arr2

Author: As long as you get rid of it in the `…` part.

>     Ur $ sum arr1 + sum arr2

Reviewer 9: In the example in the proposal (presumably) only `free` consumes `Read` and would prohibit `sum`, since (presumably) `write` gives you the `Write` constraint back much like read so you can do it again.

paper_, our solving algorithm is based on `Efficient resource
management for linear logic proof search`_, where such higher-order
givens are handled. The extension is unproblematic, it would
presumably be more effort to prevent it than to support it.

Therefore, when ``-XLinearTypes`` is
on, contexts can contain implications of the form ``C %1 => D``.

Implicit parameters
^^^^^^^^^^^^^^^^^^^

Referencing an implicit parameter is linear in the implicit
parameters. This means that linear implicit parameters can effectively
be used in programs

::

   foo :: (?x :: A) %1 => A
   foo x = ?x

Note that, because implicit parameters are currently implemented as
single-method type classes, this comes for free in the implementation.

Existential types proposal
^^^^^^^^^^^^^^^^^^^^^^^^^^

The `Existential types proposal`_, if they ever materialise (🤞), will
make this proposal even better. Using linear constraints in APIs most
often require returning constraints too. For that we've been using
GADTs. This was our simple mutable array API:

::

  read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a

  free :: (Read n, Write n) %1 => MArray a n -> ()

  data c /\ a where

>     Box :: c %1 => a -> c /\ a

Author: It isn't. I do admit to being cheeky here. But because in my examples the only linear things are constraints, it removes some bureaucratie from the proposal.

  new :: Linearly %1 => Int -> NewMArray a

  data NewMArray a where
    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a

With this API we can write functions such as

::

  read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
  read2AndDiscard arr = DataFlow.do
       (Box x) <- read arr 0
       (Box y) <- read arr 1
       free arr
       in (x, y)

  linearly $ DataFlow.do
    (NewMArray arr1) <- new
    (NewMArray arr2) <- new
    … -- modify the array as suited
    Ur $ sum arr1 + sum arr2

There is still a little bit of noise there, what with the defining of

> GADTs (while ``/\`` can be defined once and for all, types like

Author: I did intend `/\` to be used with linear constraints only here. It's not tremendously useful with non-linear constraints anyway (unless you have primitive existential types).

very strongly thematically related to linear types, the
``-XLinearTypes`` extension is still evolving anyway, and the change
is fully backward compatible.

Unrestricted modifier for constraints
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We could have an equivalent of ``Ur`` for constraint (let's write it
``UrC`` in this section). That is

::

   UrC C %1 => T  ≈ C => T

It doesn't seem quite useful at this point. Instead of

::

   (C, UrC D) %1 => T

We can write

::

   C %1 => D => T

It doesn't make a lot of difference. We may find out, with practice,
that having ``UrC`` would be preferable, but specifying it today seems
premature. Note that because of the limitations on the constraint
solver (specifically Lemma 5.5 from the paper_), ``UrC`` can't be
defined in user-land, it would need to be a specially understood
constructor.

In practice ``Ur`` is most useful when returning values, rather than
taking them as an argument. But in this case we are packaging
constraints in data types, and its easy to require them to be
unrestricted (if it ever shows up, since returning an unrestricted
constraint a rather niche concern):

> 

Author: Yes, but let's cross that bridge when we get there.

``C`` will be used. However, as we are in a linear context, we must
also ensure that it isn't *overused*:

::

   overusing :: C %1 => (Int, Int)
   overusing = (useC, useC)

Since ``overusing`` consumes a linear ``C`` constraint twice, it is
rejected for violating linearity. We may amend the type scheme of
``overusing`` in the following way, however:

::

   stammering :: (C, C) %1 => (Int, Int)
   stammering = (useC, useC)

By providing an additional linear constraint, the resource usage
guarantees are satisfied and the program type checks.

The theoretical basis of this proposal, as well as further examples,
can be found in the `Linearly Qualified Types paper <paper_>`_ by
Spiwack et al. You may also want to watch the introductory
`talk at ICFP 2022 <talk_>`_ [17min].

Motivation
----------

The motivation of this proposal is to remove unnecessary boilerplate
that is often introduced when programming with linear types.

Threading
^^^^^^^^^

Consider the following program:

::

   read2AndDiscard :: MArray a %1 -> (Ur a, Ur a)
   read2AndDiscard arr0 =
     let (arr1, x) = read arr0 0
         (arr2, y) = read arr1 1
         () = free arr2

>      in (x, y)

Author: You're not really engaging with my question if you allow yourself to arbitrarily rewrite the expression under consideration. Though I'll concede than on this particular example, without existential types, making it friendly to your syntax is not worse to making it work with linear constraints (existential types would make this example really good with linear constraints).

But you're proposal is quite fragile. What if you return a different number of arrays than are arguments (more returned: slice an array; fewer returned: write an array into another)? What if they need to be returned in something else than a tuple (because of some shared existential variable maybe?)? What if the number of arrays is decided algorithmically (that is the return type is an `Either`, it happens to sockets)? What if the argument isn't a variable (_e.g._ a function applied to a variable)? To be fully transparent, we, the co-authors of the linear types paper, have considered, and rejected pretty much this same idea back in 2017. You'll understand that I'm not very enthusiastic at the idea of relitigating it. Especially since it only addresses a subset of the motivating examples (of which there are, beyond the text of this proposal, several pages in the paper and two 2000-word blog posts). In fact, I don't even understand why we're having this conversation.

two elements. The fact that the array is a linear resource provides
some nice guarantees which ensure memory safety. For example, there is
no way we can go on to erroneously read from the array after
de-allocating it. Instead, when ``arr0`` is used as an argument to
``read``, ``read`` consumes it and returns the array under a new name
``arr1``. This process then repeats before finally ``free``
de-allocates the array.

While this code ensures that we use the ``MArray`` in a memory-safe
way, this re-naming process introduces boilerplate and becomes
cumbersome to both write and read. Compare to the doing the same with
the ``ST`` monad instead:

::

   read2AndDiscardST :: MArray s a -> ST s (a, a)
   read2andDiscardST arr = do
     x <- read arr 0
     y <- read arr 1
     return (x, y)

There is much less noise. To be honest, this is all a little silly:
the name management is completely mechanical; it should be handled
automatically by the compiler. This is what this proposal lets us do:
get pretty close to the latter style, without having to introduce a
monad.

A more interesting, though more complex, example involving freezing

> nested mutable structure is elaborated in `this blog post <blog_freeze_>`_.

Author: > It also appears that the capability mechanism introduced in Part 1 of the blog post is a nice secondary application of Linear Constraints, but I think there are simpler designs to solve the `freeze` problem that work today; for example, `freeze :: MArray a %1 -> ROArray a` and then overload `read` with a type class so that it works on both `MArray` and `ROArray`. (Internally, `ROArray` and `MArray` share the same representation of course).

If the elements of the array is unrestricted, yes, this is what we do today in linear-base. But if the elements of the array are themselves linear, they must be themselves frozen recursively, and this scheme won't let you. But you don't want to traverse the array for that. This is the $O(1)$ nested freeze problem.

Reviewer 10: Nested freeze is a good point. (Although the GC might cry out in pain for not properly $O(n)$ freezing nested arrays, instead $O(1)$ `unsafeCoerce`ing them.) Perhaps you can distill this insight into an example for the motivation?

Author: I took the opportunity of https://github.com/ghc-proposals/ghc-proposals/pull/621#discussion_r1952517690 and tried something, does it help?

Author: (sorry for the many emails, I obviously missed things when I tried to answer all in one go. I'm paranoid now)

> Although the GC might cry out in pain for not properly O ( n ) freezing nested arrays, instead O ( 1 ) `unsafeCoerce`ing them

Can you elaborate? This appears to be beyond my knowledge of the GC. I was under the (mistaken?) believe that literally nothing happened outside of the type system when you froze an array.

Reviewer 10: Mutable things potentially have to be traversed during every GC minor collection, even if they live in the old generation. I don't know enough about the details either; I just found the following code lowering `unsafeFreezeArray#`:

```
{- Freezing arrays-of-ptrs requires changing an info table, for the
   benefit of the generational collector.  It needs to scavenge mutable
   objects, even if they are in old space.  When they become immutable,
   they can be removed from this scavenge list.  -}

--  #define unsafeFreezzeArrayzh(r,a)
--      {
--        SET_INFO((StgClosure *)a,&stg_MUT_ARR_PTRS_FROZEN_DIRTY_info);
--        r = a;
--      }
  UnsafeFreezeArrayOp -> \[arg] -> opIntoRegs $ \[res] ->
    emit $ catAGraphs
      [ setInfo arg (CmmLit (CmmLabel mkMAP_FROZEN_DIRTY_infoLabel)),
        mkAssign (CmmLocal res) arg ]
```

Usually, mutable `Array#` has a "card table" that tracks which entries have been modified. It is not unlikely that GC resets this card table, so the difference between a mutable array with reset card table and an immutable array shouldn't be too large (I hope?). I fear that the same could not be said for `SmallArray#`, which lacks such a card table. Perhaps @bgamari knows whether it is a good idea to omit nested `unsafeFreeze`s on arrays, as outlined in the [blog post](https://www.tweag.io/blog/2023-01-26-linear-constraints-freeze/)? 

---

I'm happy with the new example based on O(1) `freeze` (at least in principle); I cannot see how to achieve the same without a solution based on linear constraints. 

Though you might be a bit more cautious saying that "this is an O(1) function which simply changes the type of the array", because `unsafeFreeze` is actually not a no-op... It mutates the heap object's tag, after all. I'm fine with omitting details about "well, actually omitting the freeze puts additional strain on the GC".

Author: Damn! I was _absolutely convinced_ that I knew for a fact that `unsafeFreeze` was merely `unsafeCoerce` in disguise. This is certainly challenging…

Also, I thought you knew precisely what to scavenge due to the write barrier. GHC has a write barrier, right?

It goes to show that the things you think you know… @bgamari I think I need much teaching. I'll try and come to the GHC call tomorrow, hopefully we'll have time for a little bit of that.

> because `unsafeFreeze` is actually not a no-op... It mutates the heap object's tag, after all

Just the info table right?

It does seem as if it's “just” an optimisation for the GC. Yet we want this optimisation to come through. Mutating the info table recursively would be much better than copying all the arrays, but still a long shot from O(1). An ideal solution would be if the GC itself could realise that the array has become immutable and mutate the info table itself. Myeah… needs thought.

> Will have a look at your other replies soon-ish, but lack of an example that motivated me was my biggest gripe with the proposal.

Mmm… it was never meant to be read independently of the paper and blog posts, because I thought it was too hard to write a proposal which could be read independently (it's already a pretty hefty 9000 words long). But there's always room for improvement, and I'm happy to improve. Maybe the most efficient would be to have a call and go over what you feel needs strengthening.

Reviewer 10: Re: `unsafeFreeze`: I feel like the runtime system does not actually need to differentiate `MutableArray#` vs. `Array#`. It only needs to know whether some array had been mutated since the last GC, but for `Array#` this information is stored in the card table. Therefore, `Array#` could be made indistinguishable at runtime from `MutableArray#`. There are just some recently mutated `MutableArray#`s as well as recently frozen `Array#`s with a non-reset card table. After the next GC, all card tables should be reset, and a `MutableArray#` that is never mutated again should behave exactly the same as an immutable `Array#`. In this world, `unsafeFreeze` is indeed a no-op. I don't see why it cannot become a reality, but Ben might know a good reason against.

The same is true for `SmallArray#`, where the info table pointer (DIRTY variant or not) encodes the information necessary for the GC (DIRTY=need to scavenge). 

> hard to write a proposal which could be read independently

Indeed, that's tough. But a good, minimal, indisputable example helps (even if it takes some time to turn a 2000 word blog post into a paragraph). 

> Maybe the most efficient would be to have a call and go over what you feel needs strengthening.

No, I think I'm good now. I am convinced that linear constraints make sense, you have strengthened the motivation with IMO better examples and on some of the design details we just disagree. Since I have no stake in -XLinearTypes and because there apparently are users that feel pretty enthusiastic about the proposal, I won't vote against it, though.

linear constraints, this can become

::

   data NewSocket where
     NewSocket :: Unbound s %1 => Socket s -> NewSocket

   socket :: IO NewSocket
   bind :: Unbound s %1 => Socket s -> SocketAddress -> IO (Bound s /\ ())
   listen :: Bound s %1 => Socket s -> IO (Listening /\ ())
   ...

   do
     NewSocket s <- socket -- Unbound s
     Box () <- bind s (SocketAddressInet6 S.inet6Any 8080 0 0) -- Bound s
     Box () <- listen s -- Listening s
     ...

This modifies the program much less. The source of the ``s`` variable
is more explicit. And with a little bit of ``QualifiedDo`` (or
``RebindableSyntax``) work, we can modify the program even less from
the idiomatic form (it would be even better with primitive existential
type, see the corresponding section below).

::

   Typestate.do
     NewSocket s <- socket
     bind s
     listen s
     ...

Creating linear values
^^^^^^^^^^^^^^^^^^^^^^

A well documented difficulty, when writing APIs for mutable data as
above, is that in order to guarantee that, say, an array is unique, it isn't
sufficient that ``read`` and ``write`` be linear functions. If I
create an array with

::

>    new :: Int -> MArray -- or Int %1 -> MArray

Author: > So perhaps example 1.3 that I dismissed earlier is the _main_ motivating example for this proposal? Part 2 of the blog post suggests so. 

It isn't, not particularly. In fact, it's the part of the proposal that I've explicitly marked as optional (I would really to get it in, because I find it super useful, but it's harder to design).

> However, you do not introduce the "solution", `Linearly %1 =>` to replace manual duplication of a token, until section 3.4 which is _after_ the change specification. I see that Section 3.2 of the paper does a great job of introducing `linearly` to name a region. Why don't you copy the examples and explain?

I don't quite understand your suggestion.

Solving the problems raised in the motivation using the specification in the Example and Effects section is how these proposal go. In this case the dupable constraints of Section 2.5. But, presumably, I should enhance this motivation subsection with what the solution will look like?

> Also I wonder: Can I nest multiple `linearly`s effectively, that is, can I use stuff from the outer region inside the scope of the inner region?
> 
> ```
> linearly $ 
>   let arr = newMArray 10 in
>   let arr1 = write arr 0 42 in
>   let Ur arr3 = linearly $ 
>     let pref = newMarray 2 in
>     let (arr2, Ur x) = read arr1 0 in
>     let pref2 = write pref 0 x in
>     Ur arr2 in
>   ... carry on with arr3 ...
> ```
>
> I suppose this doesn't work. Is _this_ the borrowing problem/sticky scopes you were alluding to? I.e., you want to borrow `arr1` to the inner block here, but you can't get it back afterwards. Seems like quite a fundamental limitation of `linearly`.

It's not going to work well, because the linear capabilities can't escape from the inner `linearly` scope. So you'll have to `free`/`freeze` `arr1` and `arr2` before returning from that scope. This is the sticky scopes problem (but it's mostly unrelated to the borrowing problem).

But you never need to do this. The `Linearly` constraint is duplicated appropriately by the typechecker. So a single `Linearly` is enough. You're examples rewrite to

```haskell
linearly $ 
  let arr = newMArray 10 in
  let arr1 = write arr 0 42 in
  let arr3 =
    let pref = newMarray 2 in
    let (arr2, Ur x) = read arr1 0 in
    let pref2 = write pref 0 x in
    Ur arr2 in
  ... carry on with arr3 ...
 ```

Author: > > But, presumably, I should enhance this motivation subsection with what the solution will look like?
> 
> Yes, please :) I was not able to fill in the dots.

I need to remember to do this. It got lost in my answering the thread. Grmbl!

Author: I just did (a meeting ended up a bit early, I didn't want to forget again). Does it help?

Reviewer 10: Yes, better. Now I know why 2.5 appears where it does.

non-linear constraints can then still be used in combination with
linear ones, with the order of linear and unrestricted constraints not
mattering.

Semantics/Desugaring
^^^^^^^^^^^^^^^^^^^

As typical, we define the semantics of linear constraints via
desugaring. The linear fat arrow ``%1 =>`` desugars to a linear arrow
``%1 ->``. Namely, writing ``⦇·⦈`` for the desugaring function:

- ``⦇C %1 => A⦈ = C %1 -> ⦇A⦈``
- ``⦇e :: C %1 => A⦈ = \(%1 $d :: C) -> ⦇e :: A⦈``

This desugaring means that changes to GHC Core itself are not
required: we only need the material introduced by linear types, which
is part of GHC since GHC 9.0.

Using type classes linearly
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Type class methods require an unrestricted class constraint:

::

  -- Given
  class Foo a where
    f :: F a
    g :: G a

  -- We have
  f :: Foo a => F a
  g :: Foo a => G a

This is unchanged. But we add one exception: in type classes with exactly
one method, the one method is linear in the class constraint
(otherwise there would never be inhabitants in the type ``C %1 => T``)

::

  -- Given
  class Bar a where
    h :: H a

  -- We have

>   h :: Bar a %1 => H a

Author: > Another point about the proposal I want to raise: I don't quite understand why 2.3 is necessary. Can't I just have
> 
> ```haskell
> class Bar a where
>   hinternal :: H a
> 
> h :: forall a. Bar a %1 => H a
> h = unsafeCoerce (hinternal @a)
> ```
> 
> and then do not export `hinternal`?

I must say, this is the first time that I'm asked to justify why I don't have `unsafeCoerce` in a proposal. Yes, in all our examples, linear constraints are used abstractly, so we can do any ugly unsafe coerce behind the scene. But unsafe coerce has runtime cost. So I don't see why we would implement something in GHC which can _only_ be used unsafely, just to save a handful lines of code. And it opens the door for other uses of linear constraints which I haven't designed.

> It would be sufficient if `class Read n where {}` was empty. 

No, that would be unsafe. You need to call on the constraint for real for sequencing.

> It seems awkward to me to change the way only single-method type classes are type checked, just for the reason that doing so does not impact expressivity. What is the use case to justify this complication?

I argue for why this is the way I do it in [Section 7.8](https://github.com/tweag/ghc-proposals/blob/linear-constraints/proposals/0621-linear-constraints.rst#more-methods-in-linearly-consumed-type-class). But note that, baring the change in Section 7.8 (which I don't have the capacity to deliver), even with `unsafeCoerce` you'll only be able to use single-method type class. It's the only safe coercion (there are some caveat, but they're worse, so…). Besides, there's already a special-case for single-method type classes: [WithDict](https://hackage.haskell.org/package/base-4.21.0.0/docs/GHC-Base.html#t:WithDict). Since most of the use-cases of linear constraints will involve some kind of reflection, you'd be stuck with single-method type classes even if it were safe not to.

Reviewer 10: > I don't see why we would implement something in GHC which can only be used unsafely, just to save a handful lines of code.

Whenever I can save a "handful lines of code" in the *language specification* by putting an `unsafeCoerce` in a *library* in order to encode a safe API, I would do so. It's no less safe than [using `runRW#` to call `writeArray#`](https://hackage.haskell.org/package/linear-base-0.4.0/docs/src/Data.Array.Mutable.Unlifted.Linear.html#set) in a linear API.

> But unsafe coerce has runtime cost.

It has the same runtime cost as unwrapping a `newtype`, which most of the time is zero. I would be more worried about passing around type class dictionaries; that *does* have a real cost.

> > What is the use case to justify this complication?
>
> I argue for why this is the way I do it in [Section 7.8](https://github.com/tweag/ghc-proposals/blob/linear-constraints/proposals/0621-linear-constraints.rst#more-methods-in-linearly-consumed-type-class).

You argue for why it would be much more complicated to also change the typing of multi-method type classes (I agree), but I was more interested in a use case to justify the new typing rule for single-method type classes, given that one could safely use (up to concrete evidence of the contrary) `unsafeCoerce` in that scenario (in a library such as `linear-base`).

`WithDict` is in fact a good example. It was introduced because we had *concrete evidence* that `unsafeCoerce meth :: Dict cls` can lead to miscompiles due to specialisation. We do not have similar evidence (yet) for linear constraints. If we had that evidence, altering the type-checking of single-method type classes would IMO be a convincing GHC proposal.

Reviewer 10: Let's just record then that I'm opposed to this design and would prefer not to complicate the type checker with 2.3 and encourage use of `unsafeCoerce` instead for safe library APIs. I won't vote against the proposal as written, though; perfect is the enemy of the good.

Author: Ok, I'll record this as an alternative.

difficulties we know, if you're writing such an algorithm, you don't
really have a choice but to ditch linear types and use the ``ST``
monad.

One consequence, which was a starting point of the `Linear Types
proposal`_ is that freezing a mutable array is unsafe with ``ST`` and
safe with linear types. With linear constraints, we can even
(conveniently) safely freeze nested mutable arrays (`see this blog
post <blog_freeze_>`_).

Furthermore, if you do actually want to do some manual memory
management, like we in this example, forgetting ``free`` in ``ST``
causes a hard-to-track memory leak. With linear types, it creates a
type error.

But, probably, the real motivation for using linear arrays rather than
``ST`` arrays is seen in the types. Compare

::

   read2AndDiscard :: (Read n, Write n) %1 => MArray n a -> (a, a)
   read2AndDiscard :: MArray r a -> ST r (a, a)

Plainly, and unsurprisingly, one has a monad, the other doesn't. It is
our claim (but it's presumably not a controversial claim at all) that
code without monad composes better than code with monads. The
counterargument, of course, is that the linear type ecosystem is still

> limited, and you may have to play a biggest novelty budget for linear

Reviewer 2: ```suggestion
limited, and you may have to pay a bigger novelty budget for linear
```

   listen :: HasAt n s Bound => Socket n -> IO n (Replace n s Listening) ()
   ...

Probably the hard technical bit here is to convince GHC that ``HasAt n
(Replace n s Bound) Bound`` holds. Maybe it can be done with a plugin?
Presumably with two sockets with type level names ``n`` and ``n'``,
we'd also need to be able to express that ``n`` and ``n'`` are
distinct. This looks like this API works fine when ``s`` is a concrete
list, but be quite tricky otherwise.

None of these are issues with linear constraints, because the solver
understand what it means to add a constraint (and, thanks to
linearity, to make a constraint unavailable for subsequent
computations). We don't even need to care that ``n`` and ``n'`` are
different. If we have ``(Bound n, Bound n')``, we can ``bind @n``,
then ``bind @n'`` (the API ensures that ``n`` and ``n'`` are distinct,
but it's irrelevant for the solver).

A monad as a source of uniqueness
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

An existing way of using the duplication of a comonoid implicitly is
with a (linear) reader monad.

::

   newtype Reader e a = Reader (e %1 -> a)
   instance (Dupable e) => Monad (Reader e)

> Could that replace the dupable constraint. Notably for the main

Reviewer 2: ```suggestion
Could that replace the dupable constraint? Notably for the main
```
